File: cpp/src/readwrite.cpp
#include "types.hpp"

#include "util.hpp"

#include <cstring>

namespace filefs {

std::size_t FileSystem::fread(File& f, const std::span<std::byte> out) {
  return fread(out.data(), 1U, out.size(), f);
}

std::size_t FileSystem::fread(void* ptr, const std::size_t size, const std::size_t nmemb, File& stream) {
  if (!state_ || !state_->mounted() || ptr == nullptr || !stream.open_) {
    return 0U;
  }
  if (stream.mode_ == 1U || stream.mode_ == 2U || stream.pos_blockindex_ == 0U) {
    return 0U;
  }

  const std::size_t wannasize = size * nmemb;
  std::size_t k = 0U;
  detail::Block block{};
  std::uint32_t blockindex = stream.pos_blockindex_;

  auto* out = static_cast<std::uint8_t*>(ptr);
  while (true) {
    if (!state_->readblock(blockindex, block)) {
      return 0U;
    }
    const std::uint32_t nextindex =
        detail::b4_to_u32(std::span<const std::uint8_t, 4>(block.data() + 4, 4));

    if (stream.pos_offset_ == BlockSize) {
      stream.pos_blockindex_ = blockindex;
      stream.pos_offset_ = static_cast<std::uint16_t>(detail::BlockHead);
    }

    if (blockindex == stream.file_stop_blockindex_) {
      int n = static_cast<int>(stream.file_offset_) - static_cast<int>(stream.pos_offset_);
      if (n <= 0) {
        return k;
      }
      if (wannasize - k < static_cast<std::size_t>(n)) {
        n = static_cast<int>(wannasize - k);
      }
      std::memcpy(out + k, block.data() + stream.pos_offset_, static_cast<std::size_t>(n));
      k += static_cast<std::size_t>(n);
      stream.pos_blockindex_ = blockindex;
      stream.pos_offset_ = static_cast<std::uint16_t>(stream.pos_offset_ + n);
      stream.pos_ += static_cast<std::uint64_t>(n);
      return k;
    }

    int n = static_cast<int>(BlockSize) - static_cast<int>(stream.pos_offset_);
    if (n <= 0) {
      return k;
    }
    if (wannasize - k < static_cast<std::size_t>(n)) {
      n = static_cast<int>(wannasize - k);
    }
    std::memcpy(out + k, block.data() + stream.pos_offset_, static_cast<std::size_t>(n));
    k += static_cast<std::size_t>(n);
    stream.pos_blockindex_ = blockindex;
    stream.pos_offset_ = static_cast<std::uint16_t>(stream.pos_offset_ + n);
    stream.pos_ += static_cast<std::uint64_t>(n);
    if (k >= wannasize) {
      return k;
    }

    blockindex = nextindex;
    if (nextindex == 0U) {
      return k;
    }
  }
}

std::size_t FileSystem::fwrite(File& f, const std::span<const std::byte> in) {
  return fwrite(in.data(), 1U, in.size(), f);
}

std::size_t FileSystem::fwrite(const void* ptr, const std::size_t size, const std::size_t nmemb, File& stream) {
  if (!state_ || !state_->mounted() || ptr == nullptr || !stream.open_) {
    return 0U;
  }
  if (stream.mode_ == 0U) {
    return 0U;
  }

  const std::size_t wannasize = size * nmemb;
  if (wannasize == 0U) {
    return 0U;
  }

  std::size_t cut = 0U;
  detail::Block new_block{};
  detail::Block pos_block{};
  detail::Block dir_block{};
  std::array<std::uint8_t, 4> b4{};
  std::array<std::uint8_t, 2> b2{};
  std::uint32_t next_blockindex = 0U;

  if (state_->tmp.state == 0U && !state_->tmpstart(1U)) {
    return 0U;
  }

  const auto* in = static_cast<const std::uint8_t*>(ptr);
  if (stream.pos_blockindex_ == 0U) {
    const std::uint32_t new_blockindex = state_->genblockindex();
    if (new_blockindex == 0U) {
      if (state_->tmp.state == 1U) {
        state_->tmpstop();
      }
      return 0U;
    }
    pos_block.fill(0U);
    if (!state_->writeblock(new_blockindex, pos_block)) {
      if (state_->tmp.state == 1U) {
        state_->tmpstop();
      }
      return 0U;
    }
    if (!state_->readblock(stream.dir_blockindex_, dir_block)) {
      if (state_->tmp.state == 1U) {
        state_->tmpstop();
      }
      return 0U;
    }
    detail::u32_to_b4(new_blockindex, b4);
    std::copy(b4.begin(), b4.end(), dir_block.begin() + stream.dir_offset_ - 10);
    std::copy(b4.begin(), b4.end(), dir_block.begin() + stream.dir_offset_ - 6);
    detail::u16_to_b2(static_cast<std::uint16_t>(detail::BlockHead), b2);
    std::copy(b2.begin(), b2.end(), dir_block.begin() + stream.dir_offset_ - 2);

    stream.file_start_blockindex_ = new_blockindex;
    stream.file_stop_blockindex_ = new_blockindex;
    stream.file_offset_ = 0U;
    stream.pos_blockindex_ = new_blockindex;
    stream.pos_offset_ = static_cast<std::uint16_t>(detail::BlockHead);
    stream.pos_ = 0U;
    if (!state_->writeblock(stream.dir_blockindex_, dir_block)) {
      if (state_->tmp.state == 1U) {
        state_->tmpstop();
      }
      return 0U;
    }
    next_blockindex = 0U;
  } else {
    if (!state_->readblock(stream.pos_blockindex_, pos_block)) {
      if (state_->tmp.state == 1U) {
        state_->tmpstop();
      }
      return 0U;
    }
    next_blockindex = detail::b4_to_u32(std::span<const std::uint8_t, 4>(pos_block.data() + 4, 4));
  }

  while (true) {
    if (stream.pos_offset_ == BlockSize) {
      if (next_blockindex == 0U) {
        const std::uint32_t new_blockindex = state_->genblockindex();
        if (new_blockindex == 0U) {
          if (state_->tmp.state == 1U) {
            state_->tmpstop();
          }
          return 0U;
        }
        new_block.fill(0U);
        detail::u32_to_b4(stream.pos_blockindex_, std::span<std::uint8_t, 4>(new_block.data() + 8, 4));
        detail::u32_to_b4(new_blockindex, std::span<std::uint8_t, 4>(pos_block.data() + 4, 4));
        if (!state_->writeblock(stream.pos_blockindex_, pos_block)) {
          if (state_->tmp.state == 1U) {
            state_->tmpstop();
          }
          return 0U;
        }
        stream.pos_blockindex_ = new_blockindex;
        stream.pos_offset_ = static_cast<std::uint16_t>(detail::BlockHead);
        pos_block = new_block;
        next_blockindex = 0U;
      } else {
        if (!state_->readblock(next_blockindex, pos_block)) {
          if (state_->tmp.state == 1U) {
            state_->tmpstop();
          }
          return 0U;
        }
        const std::uint32_t original_next =
            detail::b4_to_u32(std::span<const std::uint8_t, 4>(pos_block.data() + 4, 4));
        stream.pos_blockindex_ = next_blockindex;
        stream.pos_offset_ = static_cast<std::uint16_t>(detail::BlockHead);
        next_blockindex = original_next;
      }
    }

    const std::size_t avail = BlockSize - stream.pos_offset_;
    if (wannasize - cut <= avail) {
      const std::size_t n = wannasize - cut;
      std::memcpy(pos_block.data() + stream.pos_offset_, in + cut, n);
      cut += n;
      if (!state_->writeblock(stream.pos_blockindex_, pos_block)) {
        if (state_->tmp.state == 1U) {
          state_->tmpstop();
        }
        return 0U;
      }
      stream.pos_offset_ = static_cast<std::uint16_t>(stream.pos_offset_ + n);
      stream.pos_ += n;

      bool update_size = false;
      if (stream.pos_blockindex_ > stream.file_stop_blockindex_) {
        update_size = true;
      } else if (stream.pos_blockindex_ == stream.file_stop_blockindex_ &&
                 stream.pos_offset_ > stream.file_offset_) {
        update_size = true;
      }

      if (update_size) {
        if (!state_->readblock(stream.dir_blockindex_, dir_block)) {
          if (state_->tmp.state == 1U) {
            state_->tmpstop();
          }
          return 0U;
        }
        stream.file_stop_blockindex_ = stream.pos_blockindex_;
        stream.file_offset_ = stream.pos_offset_;
        detail::u32_to_b4(stream.pos_blockindex_, b4);
        std::copy(b4.begin(), b4.end(), dir_block.begin() + stream.dir_offset_ - 6);
        detail::u16_to_b2(stream.pos_offset_, b2);
        std::copy(b2.begin(), b2.end(), dir_block.begin() + stream.dir_offset_ - 2);
        if (!state_->writeblock(stream.dir_blockindex_, dir_block)) {
          if (state_->tmp.state == 1U) {
            state_->tmpstop();
          }
          return 0U;
        }
      }
      if (state_->tmp.state == 1U && !state_->commit()) {
        return 0U;
      }
      return wannasize;
    }

    const std::size_t n = avail;
    std::memcpy(pos_block.data() + stream.pos_offset_, in + cut, n);
    cut += n;
    if (!state_->writeblock(stream.pos_blockindex_, pos_block)) {
      if (state_->tmp.state == 1U) {
        state_->tmpstop();
      }
      return 0U;
    }
    stream.pos_offset_ = static_cast<std::uint16_t>(stream.pos_offset_ + n);
    stream.pos_ += n;

    if (wannasize - cut == 0U) {
      bool update_size = false;
      if (stream.pos_blockindex_ > stream.file_stop_blockindex_) {
        update_size = true;
      } else if (stream.pos_blockindex_ == stream.file_stop_blockindex_ &&
                 stream.pos_offset_ > stream.file_offset_) {
        update_size = true;
      }
      if (update_size) {
        if (!state_->readblock(stream.dir_blockindex_, dir_block)) {
          if (state_->tmp.state == 1U) {
            state_->tmpstop();
          }
          return 0U;
        }
        stream.file_stop_blockindex_ = stream.pos_blockindex_;
        stream.file_offset_ = stream.pos_offset_;
        detail::u32_to_b4(stream.pos_blockindex_, b4);
        std::copy(b4.begin(), b4.end(), dir_block.begin() + stream.dir_offset_ - 6);
        detail::u16_to_b2(stream.pos_offset_, b2);
        std::copy(b2.begin(), b2.end(), dir_block.begin() + stream.dir_offset_ - 2);
        if (!state_->writeblock(stream.dir_blockindex_, dir_block)) {
          if (state_->tmp.state == 1U) {
            state_->tmpstop();
          }
          return 0U;
        }
      }
      if (state_->tmp.state == 1U && !state_->commit()) {
        return 0U;
      }
      return wannasize;
    }
  }
}

void FileSystem::fclose(File& f) noexcept { f.open_ = false; }

}  // namespace filefs

File: zig/src/readwrite.zig
const types = @import("types.zig");
const util = @import("util.zig");
const txn = @import("txn.zig");
const block_mod = @import("block.zig");

pub fn fread(self: anytype, ptr: []u8, size: usize, nmemb: usize, stream: *types.File) usize {
    if (self.fp == null) return 0;
    if (stream.mode == 1 or stream.mode == 2) return 0;
    if (stream.pos_blockindex == 0) return 0;

    const want_size = size * nmemb;
    var k: usize = 0;
    var block = util.blockZero();
    var blockindex = stream.pos_blockindex;
    var b4: [4]u8 = undefined;

    while (true) {
        if (!block_mod.readblock(self, blockindex, block[0..])) {
            return 0;
        }
        @memcpy(b4[0..], block[4..8]);
        const nextindex = util.b4ToU32(b4[0..]);

        if (stream.pos_offset == types.BlockSize) {
            stream.pos_blockindex = blockindex;
            stream.pos_offset = types.BlockHead;
        }

        if (blockindex == stream.file_stop_blockindex) {
            var n: usize = @max(0, util.u16ToUsize(stream.file_offset) - util.u16ToUsize(stream.pos_offset));
            if (n == 0) {
                return k;
            }
            if (want_size - k < n) {
                n = want_size - k;
            }
            @memcpy(ptr[k .. k + n], block[stream.pos_offset .. util.u16ToUsize(stream.pos_offset) + n]);
            k += n;
            stream.pos_blockindex = blockindex;
            stream.pos_offset += util.usizeToU16(n);
            stream.pos += n;
            return k;
        }

        var n = types.BlockSize - util.u16ToUsize(stream.pos_offset);
        if (n == 0) {
            return k;
        }
        if (want_size - k < n) {
            n = want_size - k;
        }
        @memcpy(ptr[k .. k + n], block[stream.pos_offset .. util.u16ToUsize(stream.pos_offset) + n]);
        k += n;
        stream.pos_blockindex = blockindex;
        stream.pos_offset += util.usizeToU16(n);
        stream.pos += n;
        if (k >= want_size) {
            return k;
        }

        blockindex = nextindex;
        if (nextindex == 0) {
            return k;
        }
    }
}

pub fn fwrite(self: anytype, ptr: []const u8, size: usize, nmemb: usize, stream: *types.File) usize {
    if (self.fp == null) return 0;
    if (stream.mode == 0) return 0;

    const want_size = size * nmemb;
    if (want_size == 0) return 0;

    const auto_started = self.tmp.state == 0;
    if (auto_started and !txn.tmpstart(self, 1)) {
        return 0;
    }

    var cut: usize = 0;
    var new_block = util.blockZero();
    var pos_block = util.blockZero();
    var dir_block = util.blockZero();
    var b4: [4]u8 = undefined;
    var b2: [2]u8 = undefined;
    var next_blockindex: u32 = 0;

    if (stream.pos_blockindex == 0) {
        const new_blockindex = block_mod.genblockindex(self);
        if (new_blockindex == 0) {
            if (self.tmp.state == 1) txn.tmpstop(self);
            return 0;
        }
        if (!block_mod.writeblock(self, new_blockindex, pos_block[0..])) {
            if (self.tmp.state == 1) txn.tmpstop(self);
            return 0;
        }
        if (!block_mod.readblock(self, stream.dir_blockindex, dir_block[0..])) {
            if (self.tmp.state == 1) txn.tmpstop(self);
            return 0;
        }
        util.u32ToB4(new_blockindex, &b4);
        @memcpy(dir_block[stream.dir_offset - 10 .. stream.dir_offset - 6], b4[0..]);
        @memcpy(dir_block[stream.dir_offset - 6 .. stream.dir_offset - 2], b4[0..]);
        const off: u16 = types.BlockHead;
        util.u16ToB2(off, &b2);
        @memcpy(dir_block[stream.dir_offset - 2 .. stream.dir_offset], b2[0..]);
        stream.file_start_blockindex = new_blockindex;
        stream.file_stop_blockindex = new_blockindex;
        stream.file_offset = 0;
        stream.pos_blockindex = new_blockindex;
        stream.pos_offset = types.BlockHead;
        stream.pos = 0;
        if (!block_mod.writeblock(self, stream.dir_blockindex, dir_block[0..])) {
            if (self.tmp.state == 1) txn.tmpstop(self);
            return 0;
        }
        next_blockindex = 0;
    } else {
        if (!block_mod.readblock(self, stream.pos_blockindex, pos_block[0..])) {
            if (self.tmp.state == 1) txn.tmpstop(self);
            return 0;
        }
        @memcpy(b4[0..], pos_block[4..8]);
        next_blockindex = util.b4ToU32(b4[0..]);
    }

    while (true) {
        if (stream.pos_offset == types.BlockSize) {
            if (next_blockindex == 0) {
                const new_blockindex = block_mod.genblockindex(self);
                if (new_blockindex == 0) {
                    if (self.tmp.state == 1) txn.tmpstop(self);
                    return 0;
                }
                new_block = util.blockZero();
                util.u32ToB4(stream.pos_blockindex, &b4);
                @memcpy(new_block[8..12], b4[0..]);
                util.u32ToB4(new_blockindex, &b4);
                @memcpy(pos_block[4..8], b4[0..]);
                if (!block_mod.writeblock(self, stream.pos_blockindex, pos_block[0..])) {
                    if (self.tmp.state == 1) txn.tmpstop(self);
                    return 0;
                }
                stream.pos_blockindex = new_blockindex;
                stream.pos_offset = types.BlockHead;
                @memcpy(pos_block[0..], new_block[0..]);
                next_blockindex = 0;
            } else {
                if (!block_mod.readblock(self, next_blockindex, pos_block[0..])) {
                    if (self.tmp.state == 1) txn.tmpstop(self);
                    return 0;
                }
                @memcpy(b4[0..], pos_block[4..8]);
                next_blockindex = util.b4ToU32(b4[0..]);
                stream.pos_blockindex = next_blockindex;
                stream.pos_offset = types.BlockHead;
            }
        }

        const avail = types.BlockSize - util.u16ToUsize(stream.pos_offset);
        if (want_size - cut <= avail) {
            const n = want_size - cut;
            @memcpy(pos_block[stream.pos_offset .. util.u16ToUsize(stream.pos_offset) + n], ptr[cut .. cut + n]);
            cut += n;
            if (!block_mod.writeblock(self, stream.pos_blockindex, pos_block[0..])) {
                if (self.tmp.state == 1) txn.tmpstop(self);
                return 0;
            }
            stream.pos_offset += util.usizeToU16(n);
            stream.pos += n;

            var update_size = false;
            if (stream.pos_blockindex > stream.file_stop_blockindex) {
                update_size = true;
            } else if (stream.pos_blockindex == stream.file_stop_blockindex and stream.pos_offset > stream.file_offset) {
                update_size = true;
            }
            if (update_size) {
                if (!block_mod.readblock(self, stream.dir_blockindex, dir_block[0..])) {
                    if (self.tmp.state == 1) txn.tmpstop(self);
                    return 0;
                }
                stream.file_stop_blockindex = stream.pos_blockindex;
                stream.file_offset = stream.pos_offset;
                util.u32ToB4(stream.pos_blockindex, &b4);
                @memcpy(dir_block[stream.dir_offset - 6 .. stream.dir_offset - 2], b4[0..]);
                util.u16ToB2(stream.pos_offset, &b2);
                @memcpy(dir_block[stream.dir_offset - 2 .. stream.dir_offset], b2[0..]);
                if (!block_mod.writeblock(self, stream.dir_blockindex, dir_block[0..])) {
                    if (self.tmp.state == 1) txn.tmpstop(self);
                    return 0;
                }
            }
            if (self.tmp.state == 1 and !txn.commit(self)) {
                return 0;
            }
            return want_size;
        }

        const n = avail;
        @memcpy(pos_block[stream.pos_offset .. util.u16ToUsize(stream.pos_offset) + n], ptr[cut .. cut + n]);
        cut += n;
        if (!block_mod.writeblock(self, stream.pos_blockindex, pos_block[0..])) {
            if (self.tmp.state == 1) txn.tmpstop(self);
            return 0;
        }
        stream.pos_offset += util.usizeToU16(n);
        stream.pos += n;

        if (want_size - cut == 0) {
            var update_size = false;
            if (stream.pos_blockindex > stream.file_stop_blockindex) {
                update_size = true;
            } else if (stream.pos_blockindex == stream.file_stop_blockindex and stream.pos_offset > stream.file_offset) {
                update_size = true;
            }
            if (update_size) {
                if (!block_mod.readblock(self, stream.dir_blockindex, dir_block[0..])) {
                    if (self.tmp.state == 1) txn.tmpstop(self);
                    return 0;
                }
                stream.file_stop_blockindex = stream.pos_blockindex;
                stream.file_offset = stream.pos_offset;
                util.u32ToB4(stream.pos_blockindex, &b4);
                @memcpy(dir_block[stream.dir_offset - 6 .. stream.dir_offset - 2], b4[0..]);
                util.u16ToB2(stream.pos_offset, &b2);
                @memcpy(dir_block[stream.dir_offset - 2 .. stream.dir_offset], b2[0..]);
                if (!block_mod.writeblock(self, stream.dir_blockindex, dir_block[0..])) {
                    if (self.tmp.state == 1) txn.tmpstop(self);
                    return 0;
                }
            }
            if (self.tmp.state == 1 and !txn.commit(self)) {
                return 0;
            }
            return want_size;
        }
    }
}

pub fn fclose(self: anytype, stream: *types.File) void {
    self.allocator.destroy(stream);
}
